### PR TITLE
types:  tx_id

### PR DESCRIPTION
**Pull request by marbar3778**
## Description

Tx has become a Key Value struct. While this increases the size of a tx being stored, this mitigates some calls to application if we have already seen the tx.

Closes: #XXX

### CODE REVIEW

File: types/tx.go [hunk 1]
```diff
@@ -11,19 +11,51 @@ import (
...
+type Tx struct {
+	Key   []byte
+	Value []byte
+}
 
 // Hash computes the TMHASH hash of the wire encoded transaction.
+// Hash(key) + Hash(value)
 func (tx Tx) Hash() []byte {
-	return tmhash.Sum(tx)
+	var hash []byte
```
Comment (marbar3778) [inline types/tx.go:25]: This may be better as a hash of the value. This way we can still support get tx by hash as it currently stands.

**Comment by musalbas** [comment]
Thanks for this!
Can you expand on the design choice of why representing tx as a kv struct helps make pre-process compatible with mempool eviction?

**Comment by marbar3778** [comment]
> Can you expand on the design choice of why representing tx as a kv struct helps make pre-process compatible with mempool eviction?

With preprocess the app has the ability to modify a transaction therefore producing a new hash. Currently, transactions are stored as a hash in the mempool so when we call `Update(txs)` we will not be able to evict any transactions because the hash of the new tx is not one we have. 

By making the transaction into a KV we make the id immutable unless the application asks us to change it. I thought we could get away with not making the transaction a KV but this would overly complicate the interaction with the application. When transactions are committed in a block we will be able to pass the `tx_id`s to `Update()` in order to evict them.

**Comment by musalbas** [comment]
But I mean what would be the key and the value of a transaction, in the context of adding a pre-process phase?

**Comment by marbar3778** [comment]
> But I mean what would be the key and the value of a transaction, in the context of adding a pre-process phase?

This change is isolated from preprocess. Preprocess receives a list of txs (`types.Tx`)from the mempool, it doesn't care about the data structure, but in order to remove txs from the mempool you need to have an identifier (key) that is not the hash. 

This PR is trying to solve what happens after a block is committed, removal of the tx from the mempool.

Once this PR is done then the open PR I have for preprocess (https://github.com/marbar3778/lazyledger-core/pull/5) should not need any changes to be merged.

**Comment by musalbas** [comment]
Is the idea that when pre-process is implemented, the application can pick some key for its transaction that will not change even if the final transaction data (value) changes? Then transactions can be evicted from the mempool by key?

**Comment by marbar3778** [comment]
> Is the idea that when pre-process is implemented, the application can pick some key for its transaction that will not change even if the final transaction data (value) changes? Then transactions can be evicted from the mempool by key?

yes, that is correct.

File: types/tx_test.go [hunk 2]
```diff
@@ -134,12 +136,14 @@ func assertBadProof(t *testing.T, root []byte, bad []byte, good TxProof) {
...
 		pbProof tmproto.TxProof
 	)
+
 	err := pbProof.Unmarshal(bad)
 	if err == nil {
 		proof, err = TxProofFromProto(pbProof)
 		if err == nil {
 			err = proof.Validate(root)
 			if err == nil {
+				fmt.Println(proof.Proof.Total, good.Proof.Total)
```
Comment (liamsi) [inline types/tx_test.go:146]: t.Log instead?

File: mempool/clist_mempool_test.go [hunk 3]
```diff
@@ -186,32 +185,32 @@ func TestMempoolUpdate(t *testing.T) {
 
 	// 1. Adds valid txs to the cache
 	{
-		err := mempool.Update(1, []types.Tx{[]byte{0x01}}, abciResponses(1, abci.CodeTypeOK), nil, nil)
+		err := mempool.Update(1, []types.Tx{{Value: []byte{0x01}}}, abciResponses(1, abci.CodeTypeOK), nil, nil)
```
Comment (liamsi) [inline mempool/clist_mempool_test.go:188]: As a next step, the Update method will use the tx keys to actually update?

File: proto/tendermint/abci/types.proto [hunk 4]
```diff
@@ -307,10 +309,10 @@ message EventAttribute {
...
 // One usage is indexing transaction results.
 message TxResult {
-  int64             height = 1;
-  uint32            index  = 2;
-  bytes             tx     = 3;
-  ResponseDeliverTx result = 4 [(gogoproto.nullable) = false];
+  int64               height = 1;
+  uint32              index  = 2;
+  tendermint.types.Tx tx     = 3;
+  ResponseDeliverTx   result = 4 [(gogoproto.nullable) = false];
```
Comment (liamsi) [inline proto/tendermint/abci/types.proto:315]: It's not yet clear to me what the connection between Tx.Key and TxResult.result.tx_id is. Is there any yet? I think I'll need to checkout the branch locally for reviewing this properly.

Comment (liamsi) [reply proto/tendermint/abci/types.proto:315]: Will this be where the app can overwrite the key? And the Tx.key is what identifies the Tx when evicting? Is my understanding correct here?

File: types/tx_test.go (hunk 2 above)
Comment (marbar3778) [reply types/tx_test.go:146]: will remove, I was debugging a hash issue.

**Comment by marbar3778** [comment]
closing this for now but will keep the branch around

**Pull request closed without merging**